xds/internal/xdsclient/transport/grpctransport/grpctransport_ext_test.go

```diff
@@ -0,0 +1,93 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package grpctransport_test
+
+import (
+	"testing"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/internal/grpctest"
+	internalbootstrap "google.golang.org/grpc/internal/xds/bootstrap"
+	"google.golang.org/grpc/xds/internal/xdsclient/internal"
+	"google.golang.org/grpc/xds/internal/xdsclient/transport"
+	"google.golang.org/grpc/xds/internal/xdsclient/transport/grpctransport"
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+// Tests that the grpctransport.Builder creates a new grpc.ClientConn every time
+// Build() is called.
+func (s) TestBuild_CustomDialer(t *testing.T) {
+	// Override the dialer with a custom one.
+	customDialerCalled := false
+	origDialer := internal.GRPCNewClient
+	internal.GRPCNewClient = func(target string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+		customDialerCalled = true
+		return grpc.NewClient(target, opts...)
+	}
+	defer func() { internal.GRPCNewClient = origDialer }()
+
+	serverCfg, err := internalbootstrap.ServerConfigForTesting(internalbootstrap.ServerConfigTestingOptions{URI: "server-address"})
+	if err != nil {
+		t.Fatalf("Failed to create server config for testing: %v", err)
+	}
+
+	// Create a new transport and ensure that the custom dialer was called.
+	opts := transport.BuildOptions{ServerConfig: serverCfg}
+	builder := &grpctransport.Builder{}
+	tr, err := builder.Build(opts)
+	if err != nil {
+		t.Fatalf("Builder.Build(%+v) failed: %v", opts, err)
+	}
+	defer tr.Close()
+
+	if !customDialerCalled {
+		t.Fatalf("Builder.Build(%+v): custom dialer called = false, want true", opts)
+	}
+	customDialerCalled = false
+
+	// Reset the dialer, create a new transport and ensure that our custom
+	// dialer is no longer called.
+	internal.GRPCNewClient = origDialer
+	tr, err = builder.Build(opts)
+	if err != nil {
+		t.Fatalf("Builder.Build(%+v) failed: %v", opts, err)
+	}
+	defer tr.Close()
+
+	if customDialerCalled {
```

Comment:
shouldn't we also verify connection attempt happen as well because we are explicitly calling c.connect()?

Reply:
Are you talking about the call to `cc.Connect()` inside of `Build()`?

Verifying that a connection attempt happens would mean that we are actually testing `grpc.NewClient` and the `Connect` method `grpc.ClientConn`. This is a unit test and I'm not very inclined to do that here. Our e2e tests will anyway ensure that things work e2e.

Also, verifying that a connection attempt happens would mean that I spin up an actual management server here.

Let me know what you think about this.

Reply:
yes, c.Connect(). The docstring mentions "It establishes a gRPC client connection using the server URI" but that looks like specific to grpc transport implementation. The transport interface itself just mentions "Build creates a new xDS transport with the provided options". However, I think if we are specifically saying the grpc should establish a connection, we should test it as its an important operation of this API

Reply:
Yes, that's right. The docstring only applies to the grpc transport. 

I anyways modified it to not mention that we actually create a client connection. Instead I changed it to mention that we create a clientconn.